static/app/utils/profiling/profile/chromeTraceProfile.tsx

```diff
@@ -510,24 +514,151 @@ export function collapseSamples(profile: ChromeTrace.CpuProfile): {
+}
+
+const callFrameToFrameInfo = new Map<ChromeTrace.CallFrame, Frame>();
+function frameInfoForCallFrame(callFrame: ChromeTrace.CallFrame): Frame {
+  const frame = callFrameToFrameInfo.get(callFrame);
+  if (frame) {
+    return frame;
+  }
+
+  const name = callFrame.functionName || '(anonymous)';
+
+  return new Frame({
+    key: `${name}:${callFrame.url}:${callFrame.lineNumber}:${callFrame.columnNumber}`,
+    name,
+    file: callFrame.url,
+    line: callFrame.lineNumber,
+    column: callFrame.columnNumber,
+  });
+}
+
+// The following is taken from speedscope with minor changes to the logic - I'm adding some comments
+// to explain what is happening so it's hopefully useful to others. To build a tree from a list of samples,
+// we iterate over all samples while maintaining a stack. When each new sample is added, we find the common
+// root node of the new sample and our stack. The difference between the new stack and the previous stack is
+// the stack that we need to process. In case we encounter a common root between the two stacks, the stack frames of the new sample are
+// added to the existing stack while the non common stack frames are closed off the stack. There are some
+// special cases (engine events like gc, compile code etc) where frames are automatically added on top of the stack.
+// This is similar to how JS self profiling frame markers work. When gc is on top of the sample, we do not close
+// the samples, but instead append the gc frame to the top of the previous stack.
+function createCallTree({
+  cpuProfile,
+  samples,
+  sampleTimes,
+  frameIndex,
+}: {
+  cpuProfile: ChromeTrace.CpuProfile;
+  frameIndex: Map<number, ChromeTrace.ProfileNode>;
+  sampleTimes: NonNullable<ChromeTrace.CpuProfile['timeDeltas']>;
+  samples: NonNullable<ChromeTrace.CpuProfile['samples']>;
+}): Profile {
+  const profile = new ChromeTraceProfile(
+    cpuProfile.endTime - cpuProfile.startTime,
+    cpuProfile.startTime,
+    cpuProfile.endTime,
+    'thread',
+    'microseconds',
+    0
+  );
+  // Initialize an empty stack
+  const prevStack: ChromeTrace.ProfileNode[] = [];
+
+  for (let i = 0; i < samples.length; i++) {
+    // Samples point at the top of the stack
+    // https://github.com/v8/v8/blob/b8626ca445554b8376b5a01f651b70cb8c01b7dd/src/inspector/js_protocol.json#L1422
+    const nodeId = samples[i];
+    const stackTop = frameIndex.get(nodeId);
+
+    if (!stackTop) {
+      throw new Error(`Could not find frame ${nodeId} in frameIndex`);
+    }
+
+    // Check if we have a common node between the two stacks.
+    // Start at the top of our new stack and descend down to the child of the common node.
+    // until we find a node that is common between the two stacks. If a node is a special case
+    // node, then set the common root to the top of the previous stack.
+    let commonRoot: ChromeTrace.ProfileNode | null = stackTop || null;
+    while (commonRoot && prevStack.indexOf(commonRoot) === -1) {
```

Comment:
This isn't checking the order of the prevStack. Can this be problematic if the same frame appears multiple times in `prevStack`?